File: DailyChallenges/CPP/LeetCode_787.cpp
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
        unordered_map<int, vector<pair<int, int>>> m;
        for (auto & e: flights)
            m[e[0]].push_back({e[1], e[2]});
        vector<int> visited(n, INT_MAX); visited[src] = 0;
        queue<int> q; q.push(src);
        int cnt = 0;
        while (!q.empty() && cnt <= k){
            ++cnt;
            int sz = q.size();
            unordered_map<int, int> tm;
            for (int i=0; i<sz; ++i){
                int st = q.front(); q.pop();
                for (auto ed: m[st]){
                    int cost = visited[st] + ed.second;
                    if (cost < visited[ed.first]){
                        if (tm[ed.first] == 0 || tm[ed.first] > cost)
                            tm[ed.first] = cost;
                    }
                }
            }
            for (auto [k, v]: tm){
                visited[k] = v; q.push(k);
            }
        }
        for (auto cost: visited)
            cout << cost << " ";
        return (visited[dst] == INT_MAX) ? -1 :  visited[dst];
        
    }
};

File: DailyChallenges/Scala/LeetCode_787.scala
object Solution {
    def findCheapestPrice(n: Int, flights: Array[Array[Int]], src: Int, dst: Int, k: Int): Int = {
        val m = Array.fill[List[(Int, Int)]](n){Nil}
        for (a <- flights)
            m(a(0)) = (a(1), a(2)) :: m(a(0))
        
        val dsts = Array.fill[Int](n){Int.MaxValue}; dsts(src) = 0
        val q = scala.collection.mutable.Queue.empty[(Int, Int)]; q.enqueue((src, 0))
        
        var cnt = 0
        while (cnt <= k && q.nonEmpty){
            cnt += 1
            for (i <- 0 until q.length; (st, st_cost) = q.dequeue; (ed, cost) <- m(st) if cost+st_cost < dsts(ed)){
                dsts(ed) = cost + st_cost
                q.enqueue((ed, dsts(ed)))
            }
        }
        
        if (dsts(dst) == Int.MaxValue) -1 else dsts(dst)
    }
}
